tests/integration/olp-cpp-sdk-dataservice-read/VersionedLayerPrefetchPartitionsTest.cpp

```diff
@@ -0,0 +1,344 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+#include <gmock/gmock.h>
+
+#include <algorithm>
+#include <chrono>
+#include <memory>
+#include <numeric>
+#include <string>
+
+#include <matchers/NetworkUrlMatchers.h>
+#include <mocks/NetworkMock.h>
+#include <olp/authentication/Settings.h>
+#include <olp/core/cache/CacheSettings.h>
+#include <olp/core/client/OlpClientSettings.h>
+#include <olp/core/client/OlpClientSettingsFactory.h>
+#include <olp/dataservice/read/VersionedLayerClient.h>
+#include <olp/dataservice/read/model/Partitions.h>
+#include "HttpResponses.h"
+#include "PlatformUrlsGenerator.h"
+#include "ReadDefaultResponses.h"
+#include "ResponseGenerator.h"
+#include "VersionedLayerTestBase.h"
+
+using mockserver::ReadDefaultResponses;
+using olp::http::HttpStatusCode;
+using testing::_;
+namespace {
+
+namespace read = olp::dataservice::read;
+namespace client = olp::client;
+namespace http = olp::http;
+
+constexpr auto kTimeout = std::chrono::seconds(3);
+
+class VersionedLayerPrefetchPartitionsTest : public VersionedLayerTestBase {
+ protected:
+  std::vector<std::string> GeneratePartitionIds(size_t partitions_count) {
+    std::vector<std::string> partitions;
+    partitions.reserve(partitions_count);
+    for (auto i = 0u; i < partitions_count; i++) {
+      partitions.emplace_back(std::to_string(i));
+    }
+    return partitions;
+  }
+
+ protected:
+  const std::string layer_ = "testlayer";
+};
+
+TEST_F(VersionedLayerPrefetchPartitionsTest, PrefetchPartitions) {
+  auto partitions_count = 3u;
+  auto client =
+      read::VersionedLayerClient(kCatalogHrn, layer_, boost::none, settings_);
+  auto partitions = GeneratePartitionIds(partitions_count);
+  auto partitions_response =
+      ReadDefaultResponses::GeneratePartitionsResponse(partitions_count);
+  const auto request =
+      read::PrefetchPartitionsRequest().WithPartitionIds(partitions);
+  {
+    SCOPED_TRACE("Prefetch online");
+
+    ExpectVersionRequest();
+    ExpectQueryPartitionsRequest(partitions, partitions_response);
+    for (const auto& partition : partitions_response.GetPartitions()) {
+      ExpectBlobRequest(partition.GetDataHandle(), "data");
+    }
+
+    auto future = client.PrefetchPartitions(request).GetFuture();
+    ASSERT_NE(future.wait_for(std::chrono::seconds(kTimeout)),
+              std::future_status::timeout);
+
+    auto response = future.get();
+    ASSERT_TRUE(response.IsSuccessful());
+    const auto result = response.MoveResult();
+    ASSERT_EQ(result.GetPartitions().size(), partitions_count);
+
+    for (const auto& partition : result.GetPartitions()) {
+      ASSERT_TRUE(client.IsCached(partition));
+    }
+  }
+  {
+    SCOPED_TRACE("Prefetch cached");
```

Comment:
What is the "Prefetch cached" test doing? What does it have to do with caching? I don't see any GetData() with CacheOnly FetchOptions to also extract the data and make sure that it is usable.

Reply:
It suppose to check that we do not prefetch again online data which was prefetched before.

Reply:
Then it should be more like "Do not prefetch cached partitions twice".
But still, I am missing the proof that prefetch partitions can be queried with GetData(DataRequest).